bionic/code_references.py

```diff
@@ -0,0 +1,214 @@
+                # module using the import statement. If a user is importing
+                # modules inside a function, they probably don't want to import
+                # the module until the function execution time.
+                message = oneline(
+                    f"""
+                Bionic found a code reference in file ${code.co_filename} at line
+                ${lineno} that imports ${op.argval} module when hashing
+                ${code.co_name}. Changes inside the module will likely not be
+                detected. To detect changes inside the module, import it globally
+                instead.
+                """
+                )
+                warnings.warn(message)
+                set_tos(None)
+            elif op.opname in ["LOAD_METHOD", "LOAD_ATTR"]:
+                if isinstance(tos, str):
+                    tos += "." + op.argval
+                # TODO: Consider calling getattr only when TOS is a module.
+                # Doing so has risk of missing cases, and any missing case would
+                # be hard to detect for users. Before making this change,
+                # document all the cases that we won't be catching so that we are
+                # aware of them.
+                # TODO: Due to the same reason as why we don't import modules, we
+                # should not call `getattr` on properties. We should either add a
+                # separate check for properties, or implement the above TODO to
+                # fix this.
+                elif hasattr(tos, op.argval):
+                    tos = getattr(tos, op.argval)
+                else:
+                    refs.append(op.argval)
+            elif op.opname == "DELETE_FAST" and tos:
+                del context.varnames[op.argval]
+                tos = None
+            elif op.opname == "STORE_FAST" and tos:
+                context.varnames[op.argval] = tos
+                tos = None
+            elif op.opname == "LOAD_FAST" and op.argval in context.varnames:
+                set_tos(context.varnames[op.argval])
+            else:
+                # For all other instructions, add the current TOS as a
+                # reference.
+                set_tos(None)
+        except Exception as e:
+            # TODO: Provide a way for users to disable bytecode hashing
+            # for individual entities and add information on how to
+            # disable it in the error message.
+            message = oneline(
+                f"""
+            Bionic found a code reference in file ${code.co_filename}
+            at line ${lineno} that it cannot hash when hashing
+            ${code.co_name}. This should be impossible and is most
+            likely a bug in Bionic. Please raise a new issue at
+            https://github.com/square/bionic/issues to let us know.
+            """
+            )
+            raise CodeVersioningError(message) from e
```

Comment:
Thinking about this more: should this be a `CodeVersioningError`? If this a bug in Bionic, maybe it should be an `AssertionError`?